File: day6/day6.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <tuple>
#include <set>
#include <array>
#include <algorithm>

using namespace std;

// Directions as constant arrays
const array<pair<int, int>, 4> dirs = {{{-1, 0}, {0, 1}, {1, 0}, {0, -1}}};

void LOG(const string& file_name, const string& smth) {
    if (file_name.find("test.in") != string::npos) {
        cout << smth << ' ';
    }
}

tuple<vector<vector<char>>, pair<int, int>> load_data(const string& file_name) {
    ifstream file(file_name);
    vector<vector<char>> grid;
    pair<int, int> spawn = {0, 0};

    if (!file.is_open()) {
        cerr << "Error: Unable to open file " << file_name << endl;
        return {{}, {-1, -1}};
    }

    string line;
    int row = 0;
    while (getline(file, line)) {
        auto idx = line.find('^');
        if (idx != string::npos) {
            spawn = {row, static_cast<int>(idx)};
        }
        grid.push_back(vector<char>(line.begin(), line.end()));
        row++;
    }

    return {grid, spawn};
}

int sol1(const string& file_name) {
    auto [grid, spawn] = load_data(file_name);

    pair<int, int> pos = spawn;
    set<pair<int, int>> visited;
    int dir = 0;

    while (pos.first >= 0 && pos.first < grid.size() && pos.second >= 0 && pos.second < grid[0].size()) {
        if (grid[pos.first][pos.second] == '#') {
            pos = {pos.first - dirs[dir].first, pos.second - dirs[dir].second};
            dir = (dir + 1) % 4;
        }

        visited.insert(pos);
        pos = {pos.first + dirs[dir].first, pos.second + dirs[dir].second};
    }

    return visited.size();
}

bool check(vector<vector<char>>& grid, int r, int c, const pair<int, int>& spawn) {
  grid[r][c] = '#';
  pair<int, int> pos = spawn;
  int dir = 0;
  vector<bool> vis(1000000, false);
  int hash = 0;

  while (true) {
    if (pos.first < 0 || pos.first >= grid.size() || pos.second < 0 || pos.second >= grid[0].size()) {
      grid[r][c] = '*';
      return false;
    }

    if (grid[pos.first][pos.second] == '#') {
      pos = {pos.first - dirs[dir].first, pos.second - dirs[dir].second};
      dir = (dir + 1) % 4;
    }

    hash = (pos.second * grid[0].size() + pos.first) * 4 + dir;
    if (vis[hash]) {
      break;
    }

    vis[hash] = true;
    pos = {pos.first + dirs[dir].first, pos.second + dirs[dir].second};
  }

  grid[r][c] = '*';
  return true;
}

int sol2(const string& file_name) {
    auto [grid, spawn] = load_data(file_name);
    int ans = 0;

    for (size_t r = 0; r < grid.size(); ++r) {
        for (size_t c = 0; c < grid[r].size(); ++c) {
            if (grid[r][c] != '.') continue;

            if (check(grid, r, c, spawn)) {
                ans++;
            }
        }
    }

    return ans;
}

int main() {
    cout << "Star 1:\n"
         << " test: " << sol1("test.in") << "\n"
         << " answer: " << sol1("input.in") << "\n";
    cout << "Star 2:\n"
         << " test: " << sol2("test.in") << "\n"
         << " answer: " << sol2("input.in") << "\n";

    return 0;
}

File: day6/day6.py
from enum import Enum
import numpy as np

def LOG(file_name, smth):
  if "test.in" in file_name:
    print(smth)

def load_data(file_name):
  with open(file_name) as f:
    data = f.readlines()
  
  spawn = (0, 0)
  for r in range(len(data)):
    data[r].strip()
    idx = data[r].find('^')
    if idx != -1:
      spawn = (r, data[r].find('^'))
  
  return data, spawn

def sol1(file_name):
  grid, spawn = load_data(file_name)
  up, down, left, right = (-1, 0), (1, 0), (0, -1), (0, 1)
  dirs = [up, right, down, left]
  
  pos = spawn
  visited = set()
  dir = 0
  while pos[0] < len(grid) and pos[0] >= 0 and pos[1] < len(grid[0]) and pos[1] >= 0:
    if grid[pos[0]][pos[1]] == '#':
      pos = tuple(np.add(pos, (-dirs[dir][0], -dirs[dir][1])))
      dir = (dir + 1) % 4
    
    visited.add(pos)
    pos = tuple(np.add(pos, dirs[dir]))
  
  return len(visited)
  
    
def sol2(file_name):
  grid, spawn = load_data(file_name)
  up, down, left, right = (-1, 0), (1, 0), (0, -1), (0, 1)
  dirs = [up, right, down, left]
  
  pos = spawn
  visited = set()
  additional_obstacles = set()
  dir = 0
  while pos[0] < len(grid) and pos[0] >= 0 and pos[1] < len(grid[0]) and pos[1] >= 0:
    if grid[pos[0]][pos[1]] == '#':
      pos = tuple(np.add(pos, (-dirs[dir][0], -dirs[dir][1])))
      dir = (dir + 1) % 4
    
    if pos in visited:
      additional_obstacles.add(tuple(np.add(pos, dirs[dir])))
      
    visited.add(pos)
    pos = tuple(np.add(pos, dirs[dir]))
    
  LOG(file_name, additional_obstacles)
  if 'test.in' in file_name:
    for r in range(len(grid)):
      for c in range(len(grid[r])):
        if (r, c) in additional_obstacles and grid[r][c] == '.':
          print('0', end='')
        else:
          print(grid[r][c], end='')
        
  
  return len(additional_obstacles)
  
  
  
  

print(f"Star 1:\n test: {sol1('day6/test.in')}\n answer: {sol1('day6/input.in')}\n")
print(f"Star 2:\n test: {sol2('day6/test.in')}\n answer: {sol2('day6/input.in')}")